File: opencv-projects-python/background-subtraction-tracking.py
# import libraries
import cv2
import numpy as np

# KNN
KNN_subtractor = cv2.createBackgroundSubtractorKNN(detectShadows = True) # detectShadows=True : exclude shadow areas from the objects you detected

# MOG2
MOG2_subtractor = cv2.createBackgroundSubtractorMOG2(detectShadows = True) # exclude shadow areas from the objects you detected

# choose your subtractor
bg_subtractor=MOG2_subtractor

camera = cv2.VideoCapture("videos/helicopter3.mp4")

while True:
    ret, frame = camera.read()

    # Every frame is used both for calculating the foreground mask and for updating the background. 
    foreground_mask = bg_subtractor.apply(frame)

    # threshold if it is bigger than 240 pixel is equal to 255 if smaller pixel is equal to 0
    # create binary image , it contains only white and black pixels
    ret , treshold = cv2.threshold(foreground_mask.copy(), 120, 255,cv2.THRESH_BINARY)
    
    #  dilation expands or thickens regions of interest in an image.
    dilated = cv2.dilate(treshold,cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3,3)),iterations = 1)
    
     # find contours 
    contours, hier = cv2.findContours(dilated,cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    
    # check every contour if are exceed certain value draw bounding boxes
    for contour in contours:
        # if area exceed certain value then draw bounding boxes
        if cv2.contourArea(contour) > 50:
            (x,y,w,h) = cv2.boundingRect(contour)
            cv2.rectangle(frame, (x,y), (x+w, y+h), (255, 255, 0), 2)

    cv2.imshow("Subtractor", foreground_mask)
    cv2.imshow("threshold", treshold)
    cv2.imshow("detection", frame)
    
    if cv2.waitKey(30) & 0xff == 27:
        break
        
camera.release()
cv2.destroyAllWindows()

File: ObjectTracking-background-subtraction/background-subtraction-tracking.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

// path to video
std::string video_path="../../videos/bird.mp4";

// Create KNN background subtractor
cv::Ptr<cv::BackgroundSubtractor> KNN_subtractor = cv::createBackgroundSubtractorKNN(true);

// Create MOG2 background subtractor
cv::Ptr<cv::BackgroundSubtractor> MOG2_subtractor = cv::createBackgroundSubtractorMOG2(true);


int main() {

    // Choose your subtractor (here using MOG2)
    cv::Ptr<cv::BackgroundSubtractor> bg_subtractor = MOG2_subtractor;


    // Open the video file
    cv::VideoCapture video(video_path);
    cv::Mat frame, foreground_mask, threshold_img, dilated;

    while (true) {
        bool ret = video.read(frame);
        if (!ret) {
            std::cout << "End of video." << std::endl;
            break;
        }

        // Apply the background subtractor to get the foreground mask
        bg_subtractor->apply(frame, foreground_mask);

        // Apply threshold to create a binary image
        cv::threshold(foreground_mask, threshold_img, 120, 255, cv::THRESH_BINARY);

        // Dilate the threshold image to thicken the regions of interest
        cv::dilate(threshold_img, dilated, cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(3, 3)), cv::Point(-1, -1), 1);

        // Find contours in the dilated image
        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(dilated, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

        // Draw bounding boxes for contours that exceed a certain area threshold
        for (size_t i = 0; i < contours.size(); i++) {
            if (cv::contourArea(contours[i]) > 150) {
                cv::Rect bounding_box = cv::boundingRect(contours[i]);
                cv::rectangle(frame, bounding_box, cv::Scalar(255, 255, 0), 2);
            }
        }

        // Show the different outputs
        cv::imshow("Subtractor", foreground_mask);
        cv::imshow("Threshold", threshold_img);
        cv::imshow("Detection", frame);

        // Exit when 'ESC' is pressed
        if (cv::waitKey(30) == 27) break;
    }

    video.release();
    cv::destroyAllWindows();
    return 0;
}